src/domain/mongoose_domain_core.erl

```diff
@@ -0,0 +1,139 @@
+%% Generally, you should not call anything from this module.
+%% Use mongoose_domain_api module instead.
+-module(mongoose_domain_core).
+-include("mongoose_logger.hrl").
+
+-export([start/1, stop/0]).
+-export([start_link/1]).
+-export([get_host_type/1]).
+-export([is_locked/1]).
+
+%% API for tests
+-export([dump/0, restore/1]).
+
+%% API, used by DB module
+-export([remove_all_unlocked/0,
+         upsert_unlocked/2,
+         remove_unlocked/1]).
+
+-export([get_all_locked/0,
+         get_domains_by_host_type/1]).
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(TABLE, ?MODULE).
+
+start(Pairs) ->
+    ChildSpec =
+        {?MODULE,
+         {?MODULE, start_link, [Pairs]},
+         permanent, infinity, worker, [?MODULE]},
+    just_ok(supervisor:start_child(ejabberd_sup, ChildSpec)).
+
+%% For tests
+stop() ->
+    supervisor:terminate_child(ejabberd_sup, ?MODULE),
+    supervisor:delete_child(ejabberd_sup, ?MODULE),
+    ok.
+
+start_link(Pairs) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Pairs], []).
+
+get_host_type(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [] ->
+            {error, not_found};
+        [{_Domain, HostType, _Locked}] ->
+            {ok, HostType}
+    end.
+
+is_locked(Domain) ->
+    case ets:lookup(?TABLE, Domain) of
+        [{_Domain, _HostType, _Locked = true}] ->
+            true;
+        _ ->
+            false
+    end.
+
+remove_all_unlocked() ->
```

Comment:
What is the purpose of such interface? There must be per domain operations, group removal should not be possible. Otherwise, it's impossible to implement domain add/remove hooks.

Reply:
It is the simplest way to recover if gen_server of a DB service crashes.

Reply:
must be reworked, this is not acceptable. ETS must be a source of truth for other components, it cannot spontaneously lost all the domains.

Reply:
putting here some notes after today's discussion:

there are 2 different cases when domains DB service can potentially crash:
- During the MIM startup. before the initial filling of the ETS table (from DB) is done.
- after initial ETS filling.

for the first case it's acceptable to completely drop the core's ETS table, because we cannot use events DB table to perform an update of the ETS table with the latest changes.

for the second case, we must use events DB table.

implementation notes:
- core's gen_server must store the last applied event id.
- at the beginning, before the initial ETS filling, this event id is undefined.
- once initial ETS filling is done, event id is set for the first time.
- after that event id is updated with each and every change of the ETS.
- it must be possible to drop ETS data only if event id is not set yet. 
- after initialization of the event id, DB service (on restart) must use that stored event id to apply the latest changes to the ETS.
- also we may want to extend the cleaning interval and max age parameters for the events DB table to some higher value, e.g. 1 day

@arcusfelis please add some implementation note in the source code.